Code/Framework/AzCore/AzCore/std/ranges/ranges_algorithm.h

```diff
@@ -2001,6 +2001,193 @@ namespace AZStd::ranges
         constexpr Internal::ends_with_fn ends_with{};
     }
 
+    //! Constrained binary search algorithms
+    // ranges::lower_bound
```

Comment:
Ranges work got into this PR?
std/algorithm.h
std/ranges/ranges_algorithm.h
Tests/AZStd/RangesAlgorithmTests.cpp

Are they dependent changes?

Reply:
Yes, I needed to use `ranges::equal_range` and `ranges::upper_bound` to allow a binary search on the m_eventLoggers array
The `m_eventloggers` array contains a struct of a <event logger id, unique_ptr>([IdToEventLoggerEntry](https://github.com/o3de/o3de/pull/12004/files/660b83b1a67e1a45c3b285c0ac52f5a1fb838672#diff-2ae5b207ce36eec70e1b4d4f7ed76f7cfba7bbc4409cfa917358933e31b17693R59)) within it, which means to invoke an algorithm on it requires passing an instance of such a struct, but the problem, is that there can only be one owner of the unique pointer. 
So using normal `std::equal_range` requires passing an instance of `IdToEventLoggerEntry` to it to search for the value, but that entry would need to have a emptry unique_ptr value set for it.

The `ranges::equal_range` and `ranges::upper_bound` allows for a [projection function](https://en.cppreference.com/w/cpp/algorithm/ranges/equal_range) to be provided This is used that the code to extract the `EventLoggerId` from the `IdToEventLoggerEntry` structure to avoid needing to construct a temporaey `IdToEventLoggerEntry`.

See the logic in [FindEventLoggerImpl](https://github.com/o3de/o3de/pull/12004/files/660b83b1a67e1a45c3b285c0ac52f5a1fb838672#diff-07acd27142a47c2dc537cf4b4b927055822d9cb56df6a604a301736635190e3aR125-R128)